Implementation of a Command Line Setup

klees opened the pull request:
Dear all,

in this PR I wired up a lot of the stuff for the new setup implementation that I build and showed in the previous PRs for the setup. With this changes it is possible to install ILIAS from the command line without using the web gui.

This is no make or break situation, though, I only minimally changed the web gui and the database update mechanism. This mostly reimplements tasks that are also contained in the old setup as `Objectives` and distributes them to the components they belong to. If we merge this, we basically will have a second implementation of the setup that is mostly independent from the gui version. I will then move the gui version to the new mechanism in the next release and get rid of the old version then.

The changes here still leave some rough edges that imo mostly highlight structural problems that also exist in the gui version of the setup and show in some very brittle code there as well. You will find these mostly when it comes to executing update steps and loading component information from xmls. Both areas are far beyond what I can currently handle but imo seem to be valuable targets for
improvements in the next version of ILIAS.

I will add a [README](https://github.com/klees/ILIAS/blob/trunk_setup/setup/README.md) tomorrow, I'm tired now after fighting dragons the whole day. I invite everyone to try the new setup implementation then, and also look into the code, asks questions, etc. I expect that there will be stuff to do, even after this PR is merged, as there most certainly will be things I did not know or think about, as well as improvements for the usability of the command line setup.

I will also add comments to this PR for every maintainer of components where I changed or added code, to highlight what they should review and scrutinize. Most likely also tomorrow.

Finally, I will take this to the next JF to show and discuss it and decide if we want to include this in ILIAS 6. I think this is too important to make this decision only via this PR or (worse) unilateraly.

Until then (and also afterwards, of course) I'm very happy to hear your feedback and questions.

Best regards!

nhaagen committed:
Component/Setup: read XMLs and store config

klees committed:
Database: removed outdated TODO

klees committed:
Data: improved handling of DataSize

klees committed:
Setup: added conditions on PHP to be checked during setup

klees committed:
Setup: removed outdated check from old setup

klees committed:
Setup: added conditions to check directory permissions

klees committed:
Setup: added objective to populate ini-file

klees committed:
Setup: added configs for various services

klees committed:
Setup: minor boyscouting

klees committed:
Http/Setup: added missing http_path to config

klees committed:
Setup: added objective to fill common ini

klees committed:
Filesystem/Setup: fill filesystem settings to ilias.ini

klees committed:
Http/Setup: fill http settings to ilias.ini

klees committed:
Logging/Setup: fill logging settings to ilias.ini

klees committed:
Style/Setup: fill style settings to ilias.ini

klees committed:
VirusScanner/Setup: fill virus scanner settings to ilias.ini

klees committed:
Setup: wired new agents in cli.php

klees committed:
Setup: objective to create directory

klees committed:
Filesystem/Setup: create webdir and datadir

klees committed:
Setup: added client_id as common field

klees committed:
Setup: create client.ini.php

klees committed:
Setup: use UnachievableException in various Objective::achieve (boy scouting)

klees committed:
Database/Setup: write ini-settings

klees committed:
Language/Setup: fill ini

klees committed:
Language/Setup: added visibility to methods in ilSetupLanguage

klees committed:
Language/Setup: moved ilSetupLanguage class to its service

klees committed:
Language/Setup: factored out common objective

klees committed:
Language/Setup: install languages via setup

The management and installation of the language seems to be a good target for
some more refactoring. This is the bare minimum that needs to be done to install
the languages from the cli.

klees committed:
Administration: added factory class for ilSetting

klees committed:
Administration/Setup: added objective to initialize settings factory

klees committed:
Database: made some optional array fields explicit in ilDBAnalyzer

klees committed:
Language/Setup: added objetive to store default language in settings

klees committed:
Systemfolder/Setup: set installation information

klees committed:
Http/Setup: added proxy-settings and some cleanup

klees committed:
Http/Setup: added proxy-settings to cli-setup

klees committed:
GlobalCache/Setup: configure global cache in cli-setup

klees committed:
BackgroundTasks/Setup: added agent, config and objectives

klees committed:
Preview/Setup: added agent, config and objective

klees committed:
MediaObject/Setup: added agent, config and objective

klees committed:
PDFGeneration/Setup: added config, agent and objective

klees committed:
MathJax/Setup: added config, agent and objective

klees committed:
Utilities/Setup: added config, agent and objective

klees committed:
Setup: fix usage fault tolerance mechanism in install command

klees committed:
Style/Setup: unified agents and renamed objective for KS-data

klees committed:
UICore/Setup: moved ilCtrlStructureReader to UICore

klees committed:
Setup: use interface finder for agents

klees committed:
ActiveRecord: empty line after ?> spoils output

klees committed:
UICore: prevent notices in ilCtrlStructureReader

klees committed:
Setup: improved ux on cli and some refactoring

klees committed:
Setup: some more refactoring

klees committed:
Setup: some speedup for the app via lazyness

klees committed:
Setup some refactoring

klees committed:
Setup: gracefully handle missing configs in agent collection

klees committed:
Setup: properly process relative filenames in config reader

klees committed:
Setup: added objective to make client accessible

klees committed:
Setup: ask for proper user when running cli

klees committed:
Setup: added objective that adds preconditions

klees committed:
Setup: silence checks for permissions on directories

klees committed:
ActiveRecord: add check to prevent PHP Notice

klees committed:
Filesystem/Setup: create missing directories

klees committed:
Database/Setup: added objective to update database

klees committed:
Database: prevent PHP Notices

klees committed:
Migration: made classname for update-step 5302 unique

klees committed:
Setup: print errors in CLI-commands

klees committed:
Certificates: fixed migration, directories might not exist

klees committed:
LSO: removed php-notices from db-updates

klees committed:
Setup: include version

klees committed:
Setup: added readme for cli setup

klees commented:
Hi folks,

I added a [README](https://github.com/klees/ILIAS/blob/trunk_setup/setup/README.md) and a [minimal-config](https://github.com/klees/ILIAS/blob/trunk_setup/setup/minimal-config.json). Should nothing prevent you to try this anymore.

Best regards!

klees commented:
Dear @alex40724!

As promised, I'll try to tell every maintainer what I have done in his (sadly, its never "her" in ILIAS) component, and I decided to start with you. So here we go:

* I introduced `Module/SystemFolder/classes/Setup`, containing three new classes for the setup. The `SystemFolder` seems to take part in the setup when it comes to information about the installation, which is what I encoded in the classes. No interesting stuff here, imo, just straight forward implementation.
* I introduced a `ilSettingsFactory` and a `Setup\Objective` to build said factory in `Services\Administration`. I want to do proper dependency injection in the setup. Settings objects are used in various places during the setup process, but the current method to create them out of thin air via `new` does not account for the various internal dependencies `ilSettings` pulls via globals/service locator. The `ilSettingsFactory` wraps around this fact and presents an abstracted interface to create settings for setup related classes. I hope we can build from there and move `ilSettings` in the same direction in the future. The peek in `ilSettings` also left me with the impression that there should be database roundtrips to save here, so the current settings-machinery might be a worthwhile target for review from that perspective as well.
* I moved the `COPageDefReader` out of the `setup/classes` directory to `Services/COPage/classes/Setup`. I think this is where it really belongs.
* Like for the `SystemFolder` I introduced classes for the setup in `Services/MediaObject`. Nothing interesting here either, imo, just setting a path to ffmpeg.
* I moved the `ilCtrlStructureReader` to `Service/UICore`.
* Like for the `SystemFolder` I introduced classes for the setup in `Services/Utilities`. Nothing interesting here either, imo, just setting paths to various tools. This might also be interesting for the other "implicit maintainer" @chfsx.

So this is mostly just distributing setup stuff to the components it belongs to. I hope this is uncontroversial. The settings, on the other hand, seem to be something we should be working on in the future. I hope we can agree on the first step I took in this PR. If there are any questions, doubts or other feedback, feel free to contact me here, via mail or via skype/phone and I'm sure we can figure something out.

Thanks for your time!

klees commented:
Dear @chfsx,

your next. As for Alex, I'll try to highlight what I did in your components and why:

* The edits in `ActiveRecord` are mostly editorial and prevent PHP notices.
* The `BackgroundTasks` got a new `Setup` subfolder that contains classes for the setup mechanism. This is mostly uninteresting (imo) and is just about making the configurations for the background tasks.
* There are a lot of changes in `Services\Database` that are only about suppressing PHP notices by making possible missing fields etc. explicit, mostly by using `?? null`. This allows the command line setup with full php error-level without notices etc. I fixed them as I found them, but I'm sure there will be more similar cases when other paths through the database setup are used (e.g. for Postgres). I would suggest to fix them accordingly, when discovered. I also added some objectives for the database for storing database config and updating the database. This spawned [this abomination](https://github.com/ILIAS-eLearning/ILIAS/pull/2313/files#diff-e04f73dad0724201e2a7238bae482175R41) that wraps around the fact, that the DB-updates really don't use dependency injection. I'm sure you are aware of this and I hope we can work on this in the future. Currently I see now better way than what I proposed. In this way we at least have the dependencies that are pulled via globals in one place, maybe this helps in the future.
* Just like the `BackgroundTasks`, the `FileSystem`-service received some classes to participate in the setup process. They are about creating folders, etc., and I'm sure they'll come in handy when you improve the service in the next release.
* The `GlobalCache` received the same treatment like the `BackgroundTasks` as well, with the notable exception that I found a config-class that I could reuse. This made me happy =)
* The `Preview` service got the setup treatment as well. Also mostly uninteresting, imo.

I also want to point out that I added a `toByte`-method to `DataSize`, which you could very well use for the `File Input` in the UI-framework.

I hope you are fine with the changes and additions to your components. Of course I'm also available for questions, feedback etc. and I also added you as a reviewer.

Thanks for your attention!

chfsx commented on Services/Database/classes/Setup/class.ilDatabaseUpdatedObjective.php:
```diff
@@ -0,0 +1,91 @@
+		$common_config = $environment->getConfigFor("common");
+		$filesystem_config = $environment->getConfigFor("filesystem");
+
+		// ATTENTION: This is a total abomination. It only exists to allow the db-
```
Thanks for clarifying this, since we will have to tackle this later. did you already test a full upgrade vom e.g.5.3 to 6?

klees replied:
No, currently I only updated from 5.4 to 6. I put this on my agenda. Also, I'm thinking about invoking some admins to test this thoroughly. Maybe @corro want to have look?

chfsx replied:
Good idea, I'll have a look at it with @corro tomorrow!

chfsx commented on Services/UICore/classes/Setup/class.ilUICoreSetupAgent.php:
```diff
@@ -10,9 +10,9 @@ class ilUICoreSetupAgent implements Setup\Agent
+	public function __construct()
 	{
-		$this->ctrl_reader = $ctrl_reader;
+		$this->ctrl_reader = new \ilCtrlStructureReader();
```
I will be happy if this is saved as an array-artefact :-)

chfsx reviewed (approved):
@klees thanks a lot, this is a huge step for ILIAS Setup! I'm absolutely fine with the changes in my party. and as you have already written, the problem of DI within the update steps we have to figure out how to tackle it. For now, it's as good as you wrote it.

klees commented:
Dear @mstuder,

now to your (and in some sense also my) component:

I introduced some standard setup classes in `Services\Component` to include the service into the setup. During the setup the service is responsible for reading module and service xmls. Like with the database this process has a lot of hidden dependencies that I collected and resolved [in this ugly snippet here](718bb4a2cc4b17e122c30e1e87f6d984R63). I moved `\ilModuleReader`, `\ilServiceReader` and  `\ilObjDefReader` from the setup to the service and made them support the injection of a database dependency. When looking into this classes it became fairly obvious to me that this is a part of ILIAS that deserves some love and some overhaul. Maybe we can look into this for the next release.

Please have a look into these changes and give some feedback (ask questions, ...).

Thanks for your support!

klees commented:
Dear @chfsx,

sorry to bother you again, but did you also have a look into the VirusScanner? I thought it already moved to Ralf, but maintenance.json says it didn't. I did some (fairly unintersting) stuff there, like for e.g. `GlobalCache`.

Thanks again!

klees commented:
Dear @Amstutz,

you also have a say here (I assume) as the maintainer of the coding side of the Styles service. I added classes to include the service in the setup. It makes some settings there. I also moved the artifact builder for the Kitchen Sink around a bit. Not too interesting overall, to me at least.

Still, I ask you to have a look into the changes and will be happy to hear your opinion, answer your questions etc.

Thanks for your attention!

klees commented:
Dear @gvollbach,

you are the one taking care of the PDFGeneration service. As this service wants to know a path to PhantomJS, I included classes for the setup in it. 

I would also be very happy if you would try to execute the setup from the command line.

Please have a look and let me know what you think.

Thanks for your time!

klees commented:
Dear @fneumann,

please also have a look here, I made changes in two of your components:

* `Services/MathJax` received some standard classes to ask for the path to Latex in the setup.
* `Services/Language` did so as well. I also moved `\ilSetupLanguage` to the language service. It is mostly concerned with installing the languages and thus imo belongs to the service.

Please have a look and give me some feedback.

Thanks for your time!

gvollbach commented:
> ...I would also be very happy if you would try to execute the setup from the command line...


Hi @klees,

I would like to, but the ` composer dump-autoload`, won't work for me on this branch:

```
Script cd .. && cd .. && php setup/cli.php build-artifacts handling the post-autoload-dump event returned with an error

                                                                                                                                                                     
  [RuntimeException]                                                                                                                                                 
  Error Output: PHP Fatal error:  Cannot declare class ILIAS\Setup\CLI\InstallCommand, because the name is already in use in /home/ILIAS/local_public_html/proj  
  ekte/www/src/Setup/CLI/InstallCommand.php on line 76                                                                                                         
  PHP Stack trace:                                                                                                                                                   
  PHP   1. {main}() /home/ILIAS/local_public_html/projekte/www/setup/cli.php:0                                                                             
  PHP   2. ILIAS\Setup\CLI\App->run() /home/ILIAS/local_public_html/projekte/www/setup/cli.php:32                                                          
  PHP   3. ILIAS\Setup\CLI\App->doRun() /home/ILIAS/local_public_html/projekte/www/libs/composer/vendor/symfony/console/Application.php:145                
  PHP   4. ILIAS\Setup\CLI\App->doRunCommand() /home/ILIAS/local_public_html/projekte/www/libs/composer/vendor/symfony/console/Application.php:269         
  PHP   5. ILIAS\Setup\CLI\BuildArtifactsCommand->run() /home/ILIAS/local_public_html/projekte/www/libs/composer/vendor/symfony/console/Application.php:9  
  08                                                                                                                                                                 
  PHP   6. ILIAS\Setup\CLI\BuildArtifactsCommand->execute() /home/ILIAS/local_public_html/projekte/www/libs/composer/vendor/symfony/console/Command/Comma  
  nd.php:255                                                                                                                                                         
  PHP   7. ILIAS\Setup\CLI\BuildArtifactsCommand->getAgent() /home/ILIAS/local_public_html/projekte/www/src/Setup/CLI/BaseCommand.php:72                   
  PHP   8. {closure:/home/ILIAS/local_public_html/projekte/www/setup/cli.php:75-81}() /home/ILIAS/local_public_html/projekte/www/src/Setup/CLI/  
  BaseCommand.php:59                                                                                                                                                 
  PHP   9. Pimple\Container->offsetGet() /home/ILIAS/local_public_html/projekte/www/setup/cli.php:79                                                       
  PHP  10. {closure:/home/ILIAS/local_public_html/projekte/www/setup/cli.php:98-117}() /home/ILIAS/local_public_html/projekte/www/libs/composer  
  /vendor/pimple/pimple/src/Pimple/Container.php:113                                                                                                                 
  PHP  11. ILIAS\Setup\ImplementationOfInterfaceFinder->getMatchingClassNames() /home/ILIAS/local_public_html/projekte/www/setup/cli.php:100               
  PHP  12. ReflectionClass->__construct() /home/ILIAS/local_public_html/projekte/www/src/Setup/ImplementationOfInterfaceFinder.php:72                      
  PHP  13. spl_autoload_call() /home/ILIAS/local_public_html/projekte/www/src/Setup/ImplementationOfInterfaceFinder.php:72                                 
  PHP  14. Composer\Autoload\ClassLoader->loadClass() /home/ILIAS/local_public_html/projekte/www/src/Setup/ImplementationOfInterfaceFinder.php:72          
  PHP  15. Composer\Autoload\includeFile() /home/ILIAS/local_public_html/projekte/www/libs/composer/vendor/composer/ClassLoader.php:301     
```

If I try to use the `cli.php` I get the same error message.

klees commented:
Hi @gvollbach,

this is strange. I just tried this again and for me it works flawlessly. Did you do `composer install` first, before running `dump-autoload`? Did you use a fresh clone of the repository? Maybe there are some changes locally? What PHP-version do you use?

Ladies and Gentleman, welcome to another session of "Find the difference!" =)

Thanks for looking into this...

klees commented:
Dear @smeyer-ilias,

your last, but certainly not least. I also did some changes to your components and code that you added:

* I added some pretty boring classes to `Services/Http` to set HTTP related configuration. You are listed as the "implicit maintainer" for this service, thus I ask you to have a look.
* I did the same for the Logging service, also nothing very interesting (imo) here.
* I moved and renamed a bunch of classes in `Service/Migration/DB_Update5295`. Basically I replaced `ilMD` with `ilMD5295` to deduplicate the class names. Classes with the same names are included in `Services/Metadata`. These are crawled in the beginning of the setup process when the setup looks for `Agent`s that want to participate in the setup process. Later on, the setup executes the database update steps, where your steps includes the classes in question. This causes a naming-clash which makes PHP unhappy. I thus renamed the classes and files, the the Migration service and in the db updates as well.

Feel free to ask any questions and of course give feedback.

Thanks for your effort!

gvollbach commented:
> Hi @gvollbach,
> 
> this is strange. I just tried this again and for me it works flawlessly. Did you do `composer install` first, before running `dump-autoload`? Did you use a fresh clone of the repository? Maybe there are some changes locally? What PHP-version do you use?

Yes. Yes. Fresh clone, cloned fresh again. Nope. 7.0, 7.1, 7.2 & 7.3 ;)

corro commented:
Hi @klees

Thank you for your awesome work! I'm looking forward to using the new setup script and this will certainly facilitate the life of admins, but also developers (I'm thinking of development environments).

Two things that I would like to emphasize:
* I would like to see the possibility to use environment variables as an alternative to using JSON config files. This would make automation tasks easier, especially with regards to Docker.
* IMO, the minimal config (https://github.com/klees/ILIAS/blob/trunk_setup/setup/minimal-config.json) could be shrunk even further by assuming some default values:
    * server_timezone: Assume UTC
    * database: Assume some default values, e.g. "localhost" as host
    * utilities: Assume default paths, e.g. /usr/bin/zip
    * ...

I will gladly give you more feedback when I played around with the script a bit. Unfortunately I get this error message when I try to run it inside a Docker container:
```
root@4edf25abd2a2:/var/www/html/setup# php setup.php install ../install.json 

Installing ILIAS
================

PHP Fatal error:  Uncaught TypeError: Argument 1 passed to ILIAS\Refinery\To\Transformation\NewMethodTransformation::__construct() must be an instance of ILIAS\Refinery\To\Transformation\object, instance of ILIAS\Data\Factory given, called in /var/www/html/src/Refinery/To/Group.php on line 169 and defined in /var/www/html/src/Refinery/To/Transformation/NewMethodTransformation.php:34
Stack trace:
#0 /var/www/html/src/Refinery/To/Group.php(169): ILIAS\Refinery\To\Transformation\NewMethodTransformation->__construct(Object(ILIAS\Data\Factory), 'password')
#1 /var/www/html/src/Refinery/To/Group.php(181): ILIAS\Refinery\To\Group->toNew(Array)
#2 /var/www/html/setup/classes/class.ilSetupAgent.php(62): ILIAS\Refinery\To\Group->data('password')
#3 [internal function]: ilSetupAgent->{closure}(Array)
#4 /var/www/html/src/Refinery/Custom/Transformation.php(39): call_user_func(Object(Closure), Array)
#5 /var/www/html/src/Refinery/Custom/Transformation.php(47): ILIAS\Refinery\Custom\Transformation->transform(Array)
#6 /var/www/html/src/Setu in /var/www/html/src/Refinery/To/Transformation/NewMethodTransformation.php on line 34
```

Is this an issue with the script or is just my environment (Docker with PHP 7.1) the problem?

klees commented:
Hi @corro,

thanks for your feedback! Regarding your error message: This is indeed a PHP version problem. ILIAS 6 only supports 7.2 and 7.3.

> I would like to see the possibility to use environment variables as an alternative to using JSON config files. This would make automation tasks easier, especially with regards to Docker.

My plan here is to implement a possibility to just set config values when calling the script, e.g. like `php setup\cli.php install config.json --config="database.password=root"`. This would be sufficient, right?

> IMO, the minimal config (https://github.com/klees/ILIAS/blob/trunk_setup/setup/minimal-config.json) could be shrunk even further by assuming some default values: 

Thanks! Will do. Any further suggestions?

Best regards!

corro commented:
> thanks for your feedback! Regarding your error-message: This is indeed a PHP-version problem. ILIAS 6 only supports 7.2 and 7.3.

I was not aware of this, thank you! With PHP 7.3 the error is indeed gone. The script does check for PHP >=7.3 however, so PHP 7.2 does not seem to be supported currently. Is this intentional?

> My plan here is to implement a possibility to just set config values when calling the script, e.g. like `php setup\cli.php install config.json --config="database.password=root". This would be sufficient, right?

Having an option `--config` would make things easier, yes. However, when relying on environment variables, this would lead to quite complicated calls very fast. For example just to set basic database options (assuming this is the way `--config` is used):
```
php setup\cli.php install config.json \
    --config="database.password=$ILIAS_DATABASE_PASSWORD" \
    --config="database.user=$ILIAS_DATABASE_USER" \
    --config="database.host=$ILIAS_DATABASE_HOST" \
    --config="database.database=$ILIAS_DATABASE_NAME"
```

The advantage here is, that we _can_ define the mapping between config options and environment variables. The disadvantage is, that we _have to_ define the mapping.

I would like to see a default mapping, but that comes with some disadvantages too (e.g. added complexity, priority issues, etc.).

> > IMO, the minimal config (https://github.com/klees/ILIAS/blob/trunk_setup/setup/minimal-config.json) could be shrunk even further by assuming some default values:
> 
> Thanks! Will do. Any further suggestions?

What comes to my mind right now:
* language.default_language: Assume English
* systemfolder.client.name: Assume the value of `common.client_id`

Apart from that I ran into an error when file system permissions were not set correctly. This is, of course, my fault, but the error message suggests, that there might be a missing `require`:
```
PHP Fatal error:  Uncaught Error: Class 'UnachievableException' not found in /var/www/html/setup/classes/class.ilSetupConfigStoredObjective.php:61
Stack trace:
#0 /var/www/html/src/Setup/CLI/BaseCommand.php(87): ilSetupConfigStoredObjective->achieve(Object(ILIAS\Setup\ArrayEnvironment))
#1 /var/www/html/libs/composer/vendor/symfony/console/Command/Command.php(255): ILIAS\Setup\CLI\BaseCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#2 /var/www/html/libs/composer/vendor/symfony/console/Application.php(908): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#3 /var/www/html/libs/composer/vendor/symfony/console/Application.php(269): Symfony\Component\Console\Application->doRunCommand(Object(ILIAS\Setup\CLI\InstallCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#4 /var/www/html/libs/com in /var/www/html/setup/classes/class.ilSetupConfigStoredObjective.php on line 61
```

However, even with correct file system permissions, I currently can't complete the setup. The database update aborts with the following infamous message (I'm assuming, this is one of the "rough edges" you mentioned):
```
DEAR ADMINISTRATOR !!

Please read the following instructions CAREFULLY!

-> If you are using content styles (e.g. for learning modules) style settings related
to media container have been lost when migrating from ILIAS 5.0/5.1 to ILIAS 5.2/5.3/5.4.

-> The following dbupdate step will fix this issue and set the media container properties to values
before the upgrade to ILIAS 5.2/5.3/5.4.

-> If this issue has already been fixed manually in your content styles you may want to skip
this step. If you are running ILIAS 5.2/5.3/5.4 for a longer time period you may also not want to
restore old values anymore and skip this step.
If you would like to skip this step you need to modify the file setup/sql/dbupdate_04.php
Search for 'RUN_CONTENT_STYLE_MIGRATION' (around line 25205) and follow the instructions.

=> To proceed the update process you now need to refresh the page (F5)

Mantis Bug Report: https://ilias.de/mantis/view.php?id=23299
```

Finally it seems, that the database settings in `client.ini.php` do not reflect the values in the JSON config file:

`install.json`:
```
	"database" : {
		"type" : "innodb",
		"host" : "mysql",
		"database" : "ilias",
		"user" : "ilias",
		"password" : "ilias",
		"create_database" : true
	},
```
`client.ini.php`:
```
[db]
type = "pdo-mysql-innodb"
host = "localhost"
user = "root"
pass = ""
name = "ilias"
```

gvollbach commented:
> However, even with correct file system permissions, I currently can't complete the setup. The database update aborts with the following infamous message (I'm assuming, this is one of the "rough edges" you mentioned):

You simply have to start the setup again, it is the equivalent to hit "F5" in the browser. ;)

klees commented:
FYI: We found Guidos problem. A very old composer version (1.0.? from 2016) created a strange looking entry in the classmap that pointed to the file containing `InstallCommand` a second time. We didn't investigate further, we expect users to use more recent versions of composer.

gvollbach commented:
> FYI: We found Guidos problem. A very old composer version (1.0.? from 2016) created a strange looking entry in the classmap that pointed to the file containing `InstallCommand` a second time. We didn't investigate further, we expect users to use more recent versions of composer.

The problem are these lines in `src/Setup/CLI/InstallCommand.php:`
```
			Environment::RESOURCE_ACHIEVEMENT_TRACKER => new class implements AchievementTracker {
				public function trackAchievementOf(Objective $objective) : void {}
				public function isAchieved(Objective $objective) : bool { return false; }
			}
```

klees commented:
@gvollbach This makes total sense. Anonymous classes are a rather new feature. Thanks for finding this!

klees committed:
Setup: check for PHP > 7.2 (not 7.3, thx @corro)

klees committed:
Setup: fixed namespaces of exception (thx @corro)

klees commented:
> but the error message suggests, that there might be a missing require:

There was a namespacing issue. I fixed it, thanks for finding it.

> PHP 7.2 does not seem to be supported currently. Is this intentional?

Not it wasn't. Thanks again, and fixed.

> I would like to see a default mapping, but that comes with some disadvantages too (e.g. added complexity, priority issues, etc.).

I definitely see that there is room for improvement regarding the usage of environment variables in ILIAS, but I doubt that the setup is the right place to do this long term. As far as I understand, one would in fact like to sidetrack client.ini.php and read the settings for the db-connections from environment variables in general, right? Like an alternative source for (some) configuration besides the ini-files we currently use. So making settings from env via the setup is a band aid anyway. This is definitely something we could think about.

Thanks for the further suggestions. I'll also look into the issue with the settings. I'm not sure if I will fix the issue with the aborting setup. This is something that in fact arises in db update steps that I do not own. Sadly we currently have the best practice that db steps that want to announce something use `die` to say it. I could sidetrack this somehow, but currently do not know if it is worth the effort. The db steps are on my short list for improvement anyway. What @gvollbach says is correct, though, you could just restart and will be fine.

fneumann commented:
Dear @klees

Moving the setup stuff to the component directories is generally ok for me. Will you still take care of them?

Regarding `Services/Mathjax`
The only thing the setup does is handling the ilias.ini variable `tools/latex`. This is not related to phantomjs like the code suggests, but to the ancient mimetex.cgi which has never been officially abandoned but which I really think no one uses anymore. It is still supported by a function `ilMathJax::renderMimetex()`, but only if neither Mathjax in the browser nor Mathjax on the server is activated in Administration/Third Party Software/Mathjax. I tested it once when I moved the old `ilUtil` functions to the new the MathJex service, but not afterwards. I would say it's a perfect candidate for skipping and no one will even recognize it.

Regarding `Services\Language`
Nearly all setup related files are independent from the language service with one exception: `ilSetupLanguage` now extends `ilLanguage`. Is this intended? Are there functions of ilLanguage used by the setup which are not overridden by ilSetupLanguage? The extension makes the dependency  explicit but `ilSetupLanguage` still uses its own functions for installing languages. They assume the same file format and data structures but have no connection to their pendants in `ilObjLanguage`. I have some ideas for performance improvements of language installation and update which should be supported by the setup, too. Also the format and consistency check of language files can be shared by the setup and the language service. For ILIAS 6 it's too late but for ILIAS 7 we should refactor the code that is used by the setup and service.

klees committed:
MathJax/Setup: fixed var naming

klees commented:
Dear @fneumann,

thanks for having a look here.

> Regarding Services/Mathjax [..] This is not related to phantomjs like the code suggests, 

Having "phantom" appear in `ilMathJaxSetupConfig` was some mistake when copy/pasting code from another setup folder. Didn't want to imply any connection to phantom. Fixed it in the last commit.

> I would say it's a perfect candidate for skipping and no one will even recognize it.

It can definitely be ignored in the setup, so this should be fine. If you do not expect anyone to use it anymore, maybe really put in on the agenda for removement in ILIAS 7...

> Nearly all setup related files are independent from the language service with one exception: ilSetupLanguage now extends ilLanguage. Is this intended?

`ilSetupLanguage` extends `ilLanguage` because there where places that type-hinted on `ilLanguage`. The class `ilSetupLanguage` was called `ilLanguage` as well not long ago, but this made composer unhappy when creating the classmap (and imo is a strong smell anyway). I decided to rename and extend then.

> For ILIAS 6 it's too late but for ILIAS 7 we should refactor the code that is used by the setup and service.

I would be really happy to see this happen. I also wondered about the intended relation of the `il(Setup)Language` classes and the duplicated code. I think it would be advisable to move the code in a direction where we split the management functions from the actual use. The further could than be used in the setup and the administration. The latter could imo provide a very small interface and would disallow other services to do management stuff with the languages.

> Moving the setup stuff to the component directories is generally ok for me. Will you still take care of them?

From a general perspective: Yes. I want the setup to work well, just as every other ILIAS user should. If I move the setup and require changes in the regarding classes or if we'll find general issues with the mechanisms I implemented for the setup, I'll be happy to at least have a look or even implement. Also I consider the setup maintainership to be about the general idea and direction of the ILIAS setup, so I'll try to foster a general coherence of the setup for  various components.

On the other hand, I want the setup to become a service just like any other services in ILIAS, where components just use the provided mechanisms to participate in the subsystem in question. Similar to, e.g., the event system or the language service. From this perspective I won't take care about the setup code in the components anymore. Specifically, I won't remove or fix outdated stuff from the component's setup or cleanup code in that setup as a setup maintainer. As always, I might be interested in the code in question from other perspectives, e.g. performance or general system architecture.

I understand that this might bother you from the perspective of `ilLanguage` maintainership because you just got some more work (refactoring or even maintenance of setup code) on your table. But I really do not see any other sensible way forward. As a setup maintainer, I should not know about the internals of the components that need to do setup too much. Also, components should be able to do stuff in the setup via a defined interface without having to hack into some opaque mechanism. This is where I'm trying to move the setup.

May I commit the proposed changes to your components?

Best regards!

alex40724 commented on Modules/SystemFolder/classes/Setup/class.ilInstallationInformationStoredObjective.php:
```diff
@@ -0,0 +1,67 @@
+
+use ILIAS\Setup;
+
+class ilInstallationInformationStoredObjective implements Setup\Objective {
```
Please add a comment here and in similar places that explain the purpose of the objective.

alex40724 commented on Services/Administration/classes/class.ilSettingsFactory.php:
```diff
@@ -0,0 +1,77 @@
+		}
+		finally {
+			$GLOBALS["DIC"] = $tmp_dic;
+		}
```
Couldn't we simply pass the DB to the ilSetting constructor instead?

alex40724 reviewed (changes requested):
Hi @klees,

Thank you a lot for contributing to ILIAS.

Please answer the following questions:
- [x] What is the reason behind all the DBUpdate_5295? If autoloading conflicts, couldn't we ignore the Update folder instead? 

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] I would welcome some informative comments in class headers, at least for non-trivial objective classes.
- [x] The ilSettingFactory database handling looks complicated, couldn't we pass the database to ilSettings::__construct as optional parameter instead?

All the other "moving stuff to components" (System Folder, Media, COPage) is fine for me, thanks!

kindly,
Alex

fneumann commented:
> May I commit the proposed changes to your components?

Yes, It's fine.

> If I move the setup and require changes in the regarding classes or if we'll find general issues with the mechanisms I implemented for the setup, I'll be happy to at least have a look or even implement.

For ILIAS 6 I would prefer if you take care about the setup service related stuff (Agent, Config, Objective) and I about the language service related `ilSetupLanguage`. With experience gained during the beta and stable phase of ILIAS 6 we can move the full resposibility to the language service maintainer in ILIAS 7.

klees commented:
Hi @fneumann!

> For ILIAS 6 I would prefer if you take care about the setup service related stuff (Agent, Config, Objective) and I about the language service related ilSetupLanguage. With experience gained during the beta and stable phase of ILIAS 6 we can move the full resposibility to the language service maintainer in ILIAS 7.

This is absolutely fine with me.

Thanks!

klees commented:
Dear @alex40724,

thanks a lot for your feedback.

> What is the reason behind all the DBUpdate_5295? If autoloading conflicts, couldn't we ignore the Update folder instead?

The [classmap does not contain the `Migration` folder](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-4/libs/composer/composer.json#L46), but we still need to load these classes during the db updates. At the beginning of the setup process, `Setup\ImplementationOfInterfaceFinder` is used to find all classes that implement `Setup\Agent`. The finder uses the class map and `ReflectionClass` to find matching classes. This basically loads information about all classes in the classmap. Later on, the db step includes files from `Migration` that contain classes with the same name, which makes PHP complain.

> I would welcome some informative comments in class headers, at least for non-trivial objective classes.

Sure will do. This is the one you also marked, right?

> The ilSettingFactory database handling looks complicated, couldn't we pass the database to ilSettings::__construct as optional parameter instead?

I also hoped so at first. But: The constructor calls `read`, which in turn calls `query` on the db, which then, somewhere deeper down, accesses the global ilDB again. So I used this, certainly very ugly, workaround. I hope we can make some steps forward regarding DI in ILIAS 7, though.

Best regards!

matthiaskunkel commented:
Jour Fixe, 18 NOV 2019: Richard gave a short overview of the current implementation of the command line setup and asks all related component maintainer to give feedback (if they haven't done it already).

Amstutz commented:
Hi @klees

As already expressed, this PR and the project behind has my full support and I highly appreciate it. Thx a lot for fighting those dragons for and with us.

Concerning you question @amstutz as maintainer: Thx for already taking care of implementing the interfaces. No issues there ;-).

Concerning the rest, since you put a lot of time in the development of this improvement, so it is only fair to put some time in the review. However, I have a hard time working through it completely. If you have a particular interest in some aspect, give me a note, then I can look into it.

Neat also, the list of params in the readme, some of it where new to me. Nice to have this complete list at hand now ;-).

Thx.

mstuder reviewed (approved):
as far as i can see i am not directly affected by the changes, @chfsx has already looked at our changes. so: go ahead and thank you very much!

best regards
martin

klees commented:
Hi @Amstutz!

> Neat also, the list of params in the readme, some of it where new to me. Nice to have this complete list at hand now ;-).

This will deteriorate rather quickly, I'm afraid. I'm already thinking about if and how we could make the `Agents` pass info about their config up to the app, component maintainers will otherwise just forget to update the readme, I assume.

> If you have a particular interest in some aspect, give me a note, then I can look into it.

Thanks for the offer, but I'll turn it down for this PR here. This is mostly grunt work in implementing the already proposed interfaces.

Best regards!

klees committed:
misc/Setup: added some docu in class comments

klees commented:
Hi @alex40724!

I added some docu to various `Objectives` in your services. Please ping me, if you see the need for more docu anywhere.

Best regards!

klees committed:
misc./Setup: assumed some more defaults to get an even smaller config (thx @corro)

klees commented:
Hi @corro!

I added some more defaults to make the `minimal-config.json` even smaller.

Thanks for the suggestion!

klees committed:
Setup: added update command

klees committed:
Setup: added some docu for the cli

klees committed:
Setup: some cleanup

klees committed:
Setup: enable ConfigReader to replace fields in config read from file

klees committed:
Setup: wired configuration overwrite functionality in app

klees commented:
Hi @corro!

I added a way to overwrite the configuration values in the file as discussed.

Best regards!

klees committed:
Setup: fail directly on missing confirmation from admin

klees committed:
Setup: added an option to confirm everything

klees committed:
Setup: inform user that they run the setup on their own risk

klees committed:
Setup: added GPLv3 info

klees committed:
Setup: confirm to overwrite existing installations

klees commented:
Hi everybody!

I added a note about GPLv3, the possibility to confirm all messages during the setup process, a command to update an installation and a confirmation for install if an installation already exists.

Best regards!

klees committed:
Setup: added NIC registration

klees commented:
Hi @matthiaskunkel,

I made the NIC registration work from the command line. Btw.: What does NIC stand for?

Best regards!

mjansenDatabay commented on setup/README.md:
```diff
@@ -38,6 +38,7 @@ are printed bold**, all other fields might be ommitted. A minimal example is [he
   * **client_id** is the identifier to be used for the installation 
   * **master_password** is used to identify at the web version of the setup
   * *server_timezone* where the installation resides, given as `region/city`, e.g. `Europe/Berlin`. Defaults to `UTC`.
+  * *register_nic* sends the identification number of the installation to a server of the ILIAS society together with some informatoin about the installation.
```
`information`

klees committed:
Setup: improved documentation

smeyer-ilias reviewed (approved):
Thanks Richard, 
all changes regarding my components are fine for me.

Amstutz reviewed (approved):
See comment above, I not only approve this, I highly welcome it ;-)

alex40724 reviewed (approved):
Thanks a lot!

klees committed:
Setup: do not build artifacts on update and install.

They'll be build during `composer install` which should typically be run by root
(or someone else) anyway.

klees committed:
Setup: added command to reload the control structure

klees committed:
Setup: use setup_exit in the db-updates instead of exit or die

klees committed:
Setup: ignore messages of db-updates during install

klees committed:
Setup: add option to silence db update messages during update

klees committed:
Setup: reload component definitions on update

klees committed:
Setup: do not init ini or database on update

klees committed:
Setup: let user confirm that she wants to run as root
